backend/cmm_helpers.ml

```diff
@@ -2186,21 +2271,26 @@ let generic_apply mut clos args args_type result (pos, mode) dbg =
   | [arg] ->
     bind "fun" clos (fun clos ->
         Cop
-          (Capply (result, pos), [get_field_gen mut clos 0 dbg; arg; clos], dbg))
+          ( Capply (Extended_machtype.to_machtype result, pos),
+            [get_field_gen mut clos 0 dbg; arg; clos],
+            dbg ))
   | _ -> call_caml_apply result args_type mut clos args pos mode dbg
 
 let send kind met obj args args_type result akind dbg =
   let call_met obj args args_type clos =
     (* met is never a simple expression, so it never gets turned into an
        Immutable load *)
-    generic_apply Asttypes.Mutable clos (obj :: args) (typ_val :: args_type)
+    generic_apply Asttypes.Mutable clos (obj :: args)
+      (Extended_machtype.typ_val :: args_type)
       result akind dbg
   in
   bind "obj" obj (fun obj ->
       match (kind : Lambda.meth_kind), args, args_type with
       | Self, _, _ ->
         bind "met" (lookup_label obj met dbg) (call_met obj args args_type)
       | Cached, cache :: pos :: args, _ :: _ :: args_type ->
+        let args_type = List.map Extended_machtype.to_machtype args_type in
+        let result = Extended_machtype.to_machtype result in
         call_cached_method obj met cache pos args args_type result akind dbg
```

Comment:
Here we have the same problem than with `caml_apply` but with `caml_send`, where tagged integers get converted to `Int`.